airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py

```diff
@@ -0,0 +1,27 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+
+from dataclasses import dataclass
+from typing import Any, List, Mapping
+
+import pendulum
+import requests
+from airbyte_cdk.sources.declarative.extractors.record_extractor import RecordExtractor
+from airbyte_cdk.sources.declarative.types import Record
+
+
+@dataclass
+class ZendeskChatBansRecordExtractor(RecordExtractor):
+    """
+    Unnesting nested bans: `visitor`, `ip_address`.
+    """
+
+    def extract_records(self, response: requests.Response) -> List[Record]:
+        response_data = response.json()
+        ip_address: List[Mapping[str, Any]] = response_data.get("ip_address", [])
+        visitor: List[Mapping[str, Any]] = response_data.get("visitor", [])
+        bans = ip_address + visitor
+        bans = sorted(bans, key=lambda x: pendulum.parse(x["created_at"]) if x["created_at"] else pendulum.datetime(1970, 1, 1))
```

Comment:
After we merge 2 pieces `ip_address` and `visitor`, which have basically the same structure, but resolved into 2 properties for the single response object: `{ "ip_address": {...}, "visitor": {...} }`, we sort the records additionally by the created at, to prevent the STATE collisions later on. But originally, this was added 3 years ago, and didn't change by now. Decided to leave it as it, instead of thinking what it the best approach here to solve the unsorted records. If we do have the same functionality inside the Low-Code, we can try to optimize this part for this source.

Reply:
> to prevent the STATE collisions later on

I'm slow this morning: Can you clarify this point? Which situation would generate state collision and what would be the impact?

Reply:
And why would we have filtering?

Reply:
It's not applied directly to the Zendesk situation, but i'd like to apply the best practices everywhere. If we have some sort of sorting in the underlying CDK (Low-code) methods - i think this could be removed. 

The answer to your question: 
> Why do we sort here? I see it was like that in the [previous implementation](https://github.com/airbytehq/airbyte/blob/0113c525548080f128258042d8663a415897ef8c/airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/streams.py#L271) but I'm curious if we could document why

My best guess is to just preserve the `ASC` order, after unnesting these 2 props: { "ip_address": {...}, "visitor": {...} }. For me it looks fine to have.

The regular case is when we have `semi-incremental` meaning we pull all the data from `start_date` but filterring out the records `<=` to `state`, if there is a cursor field suitable to track the STATE (updated, created, etc).

Reply:
Ok! All of this makes sense to me. Thanks for the detailed explanation! I'm good with resolving this comment. I'll let you decide if it is worth documenting this reasoning somewhere for the next dev that might jump on this one